src/plugins/explore/public/components/query_panel/utils/type_detection/index.ts

```diff
@@ -0,0 +1,192 @@
+    if (this.nlStarters.some((p) => p.test(query))) {
+      score -= 0.2;
+      reasons.push('Contains natural language starter (e.g., show, what, can, if is)');
+    }
+
+    if (score > 0 && score < 0.3) {
+      warnings.push('Low PPL score; may be ambiguous');
+    }
+
+    return { score, reason: reasons.join('; '), warnings };
+  }
+
+  private _checkKeyValue(query: string): ScoreResult {
+    let score = 0;
+    const reasons: string[] = [];
+    const warnings: string[] = [];
+
+    const quotedKvPattern = /(\w+)\s*=\s*(['"])[^'"]*\2/g; // key="value" or key='value'
+    const unquotedKvPattern = /(\w+)\s*=\s*[^"'\s]+/g; // key=value
+    const booleanOpsPattern = /\b(and|or)\b/gi; // boolean operators
+    const sourcePattern = /^\s*source\s*=\s*\S+/i; // source= at start
+
+    const quotedMatches = [...query.matchAll(quotedKvPattern)];
+    const unquotedMatches = [...query.matchAll(unquotedKvPattern)];
+    const booleanMatches = [...query.matchAll(booleanOpsPattern)];
+    const hasSourceStart = sourcePattern.test(query);
+
+    // De-duplicate overlapping matches (quoted matches are subset of unquoted)
+    const totalKvCount = new Set([
+      ...quotedMatches.map((m) => m.index),
+      ...unquotedMatches.map((m) => m.index),
+    ]).size;
+
+    if (totalKvCount > 0) {
+      score += 0.4 + (totalKvCount - 1) * 0.1;
+      reasons.push(`Detected ${totalKvCount} key=value pair(s)`);
+    }
+
+    if (booleanMatches.length > 0) {
+      score += 0.2;
+      reasons.push(
+        `Contains boolean operator(s): ${[
+          ...new Set(booleanMatches.map((m) => m[0].toLowerCase())),
+        ].join(', ')}`
+      );
+    }
+
+    if (hasSourceStart) {
+      score += 0.1;
+      reasons.push('Starts with source= pattern');
+    }
+
+    // Reduce score if NL starters or general English words are detected
+    if (this.nlStarters.some((p) => p.test(query))) {
+      score -= 0.2;
+      reasons.push('Contains natural language starter (e.g., show, what, can)');
+    }
+
+    if (/\|/.test(query)) {
+      warnings.push('Pipe character found; possibly mixed with PPL');
+    }
+
+    return {
+      score,
+      reason: reasons.join('; '),
+      warnings,
+    };
+  }
+
+  private _checkNaturalLanguage(query: string): ScoreResult {
+    let score = 0;
+    const reasons: string[] = [];
+    const warnings: string[] = [];
+
+    if (this.nlStarters.some((p) => p.test(query))) {
+      score += 0.4;
+      reasons.push('Starts with natural language pattern (e.g., show, what, can)');
+    }
+
+    const wordCount = query.split(/\s+/).length;
+    if (wordCount > 5) {
+      score += 0.1;
+      reasons.push('Long query with multiple words');
+    }
+
+    if (/[?]/.test(query)) {
+      score += 0.1;
+      reasons.push('Contains a question mark');
+    }
+
+    if (!query.includes('|') && !query.includes('=')) {
+      score += 0.2;
+      reasons.push('No PPL or key=value syntax present');
+    }
```

Comment:
Is this logic even valid? What if a value has `|` or `=`?

Reply:
I just tried this prompt on diya: `Can you write a function in js where it gets a list of manager names as an array and returns true if one of the entry = "ashwin" | "Ani"?`, it works. I think its safe to say that it will also work on OSD?